userspace/engine/rule_loader_compiler.h

```diff
@@ -31,37 +32,45 @@ namespace rule_loader
 class compiler
 {
 public:
-	/*!
-		\brief The output of a compilation.
-	*/
-	struct compile_output
-	{
-		compile_output() = default;
-		virtual ~compile_output() = default;
-		compile_output(compile_output&&) = default;
-		compile_output& operator = (compile_output&&) = default;
-		compile_output(const compile_output&) = default;
-		compile_output& operator = (const compile_output&) = default;
-
-		indexed_vector<falco_list> lists;
-		indexed_vector<falco_macro> macros;
-		indexed_vector<falco_rule> rules;
-	};
-
 	compiler() = default;
 	virtual ~compiler() = default;
 	compiler(compiler&&) = default;
 	compiler& operator = (compiler&&) = default;
 	compiler(const compiler&) = default;
 	compiler& operator = (const compiler&) = default;
 
+	// Return a new result object, suitable for passing to
+	// compile().
+        virtual std::unique_ptr<compile_output> new_compile_output();
+
 	/*!
 		\brief Compiles a list of falco rules
 	*/
 	virtual void compile(
 		configuration& cfg,
 		const collector& col,
 		compile_output& out) const;
+protected:
+	 /*!
+                \brief Compile a single condition expression,
+                including expanding macro and list references.
+
+		returns true if the condition could be compiled, and sets
+		ast_out/filter_out with the compiled filter + ast. Returns false if
+		the condition could not be compiled and should be skipped.
```

Comment:
Does it throw? If so, in which scenarios? Also, 11 arguments seem a lot and I'm wondering if there is any way this method could be further factorized.

Reply:
It throws in the same conditions that the older implementation did, namely when compiling a filter returns an error and the rule does not have allow_unknown_fields set.

I retained all the arguments just to keep it a simpler case of moving code into a method rather than a larger cleanup. If you want me to also do the cleanup to provide a more rigid interface, I can do that, but it will add additional changes to this PR, which is already getting pretty big.